### PR TITLE
Check LEB128 encoding fits in destination integer

### LINKED ISSUE #409: Bad LEB name size encoding
While running the hello_world test the following happens:
```
       bin/wasm-as test/emcc_O2_hello_world.fromasm -o a.wasm
       bin/wasm-dis a.wasm -o ab.wast
wasm-dis: ../src/wasm-binary.h:90: void wasm::LEB<unsigned int, unsigned char>::read(std::function<MiniT ()>) [T = unsigned int, MiniT = unsigned char]: Assertion `T(payload & ((typename std::make_unsigned<T>::type(1) << shift) - 1u)) == payload && "LEB payload has ignored bits"' failed.
```
After applying patch #408 to detect these cases.

Backtrace is:
```
    assertion=assertion@entry=0x491388 "T(payload & ((typename std::make_unsigned<T>::type(1) << shift) - 1u)) == payload && \"LEB payload has ignored bits\"", file=file@entry=0x491299 "../src/wasm-binary.h", line=line@entry=90, 
    function=function@entry=0x4913fc "void wasm::LEB<unsigned long, unsigned char>::read(std::function<MiniT ()>) [T = unsigned long, MiniT = unsigned char]") at assert.c:92
#3  0x00007f9353747c32 in __GI___assert_fail (
    assertion=0x491388 "T(payload & ((typename std::make_unsigned<T>::type(1) << shift) - 1u)) == payload && \"LEB payload has ignored bits\"", file=0x491299 "../src/wasm-binary.h", line=90, 
    function=0x4913fc "void wasm::LEB<unsigned long, unsigned char>::read(std::function<MiniT ()>) [T = unsigned long, MiniT = unsigned char]") at assert.c:101
#4  0x0000000000464c26 in wasm::LEB<unsigned long, unsigned char>::read(std::function<unsigned char ()>) (
    this=0x7ffd103faeb0, get=...) at ../src/wasm-binary.h:90
#5  0x0000000000462cc2 in wasm::WasmBinaryBuilder::getU64LEB (this=0x7ffd103fb328) at ../src/wasm-binary.h:1250
#6  0x000000000045efb8 in wasm::WasmBinaryBuilder::read (this=0x7ffd103fb328) at ../src/wasm-binary.h:1160
#7  0x000000000045dec0 in main (argc=4, argv=0x7ffd103fb8a8) at ../src/wasm-dis.cpp:49
```
The decoding occurs at:
```
    // read sections until the end
    while (more()) {
      auto nameSize = getU32LEB();
```
Either the encoder is wrong, or something is out of sync. 64-bit integers wouldn't help either.

Debug output:
```
Loading 'a.wasm'...
parsing binary...
== readHeader
<==
<==
getInt8: 0 (at 0)
getInt8: 97 (at 1)
getInt16: 24832 ==>
<==
getInt8: 115 (at 2)
getInt8: 109 (at 3)
getInt16: 28019 ==>
getInt32: 1836278016 ==>
<==
<==
getInt8: 11 (at 4)
getInt8: 0 (at 5)
getInt16: 11 ==>
<==
getInt8: 0 (at 6)
getInt8: 0 (at 7)
getInt16: 0 ==>
getInt32: 11 ==>
<==
getInt8: 4 (at 8)
wasm-dis: ../src/wasm-binary.h:90: void wasm::LEB<unsigned int, unsigned char>::read(std::function<MiniT ()>) [T = unsigned int, MiniT = unsigned char]: Assertion `T(payload & ((typename std::make_unsigned<T>::type(1) << shift) - 1u)) == payload && "LEB payload has ignored bits"' failed.
```

### PR DESCRIPTION
**Pull request by jfbastien**
As found by #404, the insignificant LEB128 bits were silently dropped when dealing with signed LEB values which tripped UBSAN in hello_world.

This fixes #409.

### CODE REVIEW

**Comment by jfbastien** [comment]
@kripken I think this one is now good to go, this now leads us to another exciting UBSAN failure :)

File: src/wasm-binary.h [hunk 1]
```diff
@@ -76,24 +79,37 @@ struct LEB {
...
-      if (!(byte & 128)) break;
+      bool last = !(byte & 128);
+      T payload = byte & 127;
+      typedef typename std::make_unsigned<T>::type mask_type;
+      auto shift_mask = 0 == shift
+                            ? ~mask_type(0)
+                            : ((mask_type(1) << (sizeof(T) * 8 - shift)) - 1u);
+      T significant_payload = payload & shift_mask;
+      if (significant_payload != payload) {
+        assert(std::is_signed<T>::value && last &&
```
Comment (binji) [inline src/wasm-binary.h]: Specifically, it is required to be a sign-extension of the value's sign bit, up to but not including the high bit of the final byte. So for a 32-bit value, there are 7 * 5 == 35 bits of data possible, but only 32 are valid. So the most significant byte of an encoded 5-byte LEB128 must be either `0b00000xxx` or `0b01111xxx`. If the LEB128 value is encoded in fewer than 5 bytes, then the sign bit is always `0b0100000`, so no sign-extension is necessary.

Comment (jfbastien) [inline src/wasm-binary.h]: Right, my check isn't full-paranoid :)

File: src/wasm-binary.h [hunk 2]
```diff
@@ -76,24 +79,37 @@ struct LEB {
...
     if (std::is_signed<T>::value) {
       shift += 7;
-      if (byte & 64 && size_t(shift) < 8*sizeof(T)) {
-        size_t sext_bits = 8*sizeof(T) - size_t(shift);
+      if ((byte & 64) && size_t(shift) < 8 * sizeof(T)) {
+        size_t sext_bits = 8 * sizeof(T) - size_t(shift);
         value <<= sext_bits;
         value >>= sext_bits;
-        assert(value < 0);
+        assert(value < 0 && "signe-extend should produces a negative value");
```
Comment (kripken) [inline src/wasm-binary.h]: typo `signe`

File: src/wasm-binary.h [hunk 3]
```diff
@@ -76,24 +79,37 @@ struct LEB {
...
     value = 0;
     T shift = 0;
     MiniT byte;
     while (1) {
       byte = get();
-      value |= ((T(byte & 127)) << shift);
-      if (!(byte & 128)) break;
+      bool last = !(byte & 128);
+      T payload = byte & 127;
+      typedef typename std::make_unsigned<T>::type mask_type;
```
Comment (kripken) [inline src/wasm-binary.h]: should be `maskType`, or alternatively we could rewrite this in python ;)

Comment (kripken) [inline src/wasm-binary.h]: although I see this error exists later down in the code too... so can leave for now i guess

**Comment by kripken** [comment]
lgtm but one typo

**Pull request merged**